File: contract-lib/src/main/scala/fssi/contract/States.scala
package fssi.contract

import _root_.java.util
import _root_.scala.collection.JavaConverters._

case class States(states: Map[String, AccountState]) {
  def of(accountId: String): Option[AccountState] = states.get(accountId)
  def update(accountState: AccountState): States = States(states + (accountState.accountId -> accountState))

  def updateStates(otherStates: States): States = {
    val m = otherStates.states.filter {
      case (key, _) => states.contains(key)
    }
    States(states ++ m)
  }

  lazy val bytes: Array[Byte] = {
    states.toVector.sortBy(_._1).foldLeft(Array.emptyByteArray) {(acc, n) =>
      acc ++ n._1.getBytes("utf-8") ++ n._2.bytes
    }
  }
}

object States {
  def empty: States = States(Map.empty[String, AccountState])

  def apply(jstates: Seq[AccountState]): States = States(jstates.map(x => x.accountId -> x).toMap)
  def apply(jstates: util.List[AccountState]): States = apply(jstates.asScala)
}

File: contract-lib/src/main/java/fssi/contract/java/States.java
package fssi.contract.java;

import fssi.contract.States$;
import scala.Option;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

public final class States {
    private fssi.contract.States inner;

    private States() {}

    public static  States fromScala(fssi.contract.States scalaStates) {
        States s = new States();
        s.inner = scalaStates;
        return s;
    }

    public fssi.contract.States asScala() {
        return inner;
    }

    public States(AccountState ... accountStates) {
        List<AccountState> states = Arrays.asList(accountStates);
        List<fssi.contract.AccountState> tmp = new ArrayList<>();
        for(AccountState as: states) {
            tmp.add(as.toScala());
        }
        inner = States$.MODULE$.apply(tmp);
    }

    public Optional<AccountState> of(String accountId) {
        if(accountId == null) return Optional.empty();

        Option<fssi.contract.AccountState> as = inner.of(accountId);
        if(as.isDefined()) {
            return Optional.of(AccountState.fromScala(as.get()));
        }
        return Optional.empty();
    }

    public States update(AccountState accountState) {
        if(accountState != null) {
            this.inner = inner.update(accountState.toScala());
        }
        return this;
    }
}
